File: dsa/linkedlist/reverseLinkedList.js
/**
 * Definition for singly-linked list.
 */
export class ListNode {
	/**
	 * Creates a new ListNode.
	 * @param {number} value The value of the node
	 * @param {ListNode} next The reference to the next node (default is null)
	 */
	constructor(value = 0, next = null) {
		this.value = value;
		this.next = next;
	}
}

/**
 * This function reverses a singly linked list.
 * @param {ListNode} head The head of the linked list
 * @returns {ListNode} The new head of the reversed linked list
 * */
export function reverseLinkedList(head) {
	let prev = null;
	let current = head;
	while (current) {
		const nextNode = current.next; // Store next node
		current.next = prev;           // Reverse the link
		prev = current;                // Move prev to current
		current = nextNode;            // Move to next node
	}
	return prev; // New head of the reversed list
}

/**
 * Helper function to print the linked list (for testing purposes).
 * @param {ListNode} head The head of the linked list
 */
export function printList(head) {
	let current = head;
	const values = [];
	while (current) {
		values.push(current.value);
		current = current.next;
	}
	console.log(values.join(' -> '));
}

// tiny self-checks in console
if (typeof window !== "undefined")
(function selfTest() {
	const head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
	console.log('Original list:');
	printList(head);
	console.log('---');
	const reversedHead = reverseLinkedList(head);
	console.log('Reversed list:');
	printList(reversedHead);
	console.log('---');
})();

File: dsa_in_java/linkedlist/ReverseLinkedList.java
/*
 * Topic: Reverse a Linked List
 * Language: Java
 *
 * Description:
 * Given the head of a singly linked list, reverse the list and return the new head.
 * 
 * Example:
 * Input: 1 → 2 → 3 → 4 → 5 → NULL
 * Output: 5 → 4 → 3 → 2 → 1 → NULL
 *
 * Approaches:
 * Iterative Approach:
 *    - Use three pointers (prev, current, next).
 *    - Reverse the direction of each node’s next pointer.
 *    - Time: O(n), Space: O(1)
 *
 * Recursive Approach:
 *    - Recur until the end, then reverse links while returning.
 *    - Time: O(n), Space: O(n) (stack space)
 *
 * LeetCode Link:
 * https://leetcode.com/problems/reverse-linked-list/
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

public class ReverseLinkedList {

    /**
     * Reverses the linked list using an iterative approach.
     * @param head The head of the linked list
     * @return The new head after reversal
     */
    public static ListNode reverseIterative(ListNode head) {
        ListNode prev = null;
        ListNode current = head;

        while (current != null) {
            ListNode nextNode = current.next; // store next
            current.next = prev;              // reverse link
            prev = current;                   // move prev forward
            current = nextNode;               // move current forward
        }
        return prev; // prev becomes new head
    }

    /**
     * Reverses the linked list using recursion.
     * @param head The head of the linked list
     * @return The new head after reversal
     */
    public static ListNode reverseRecursive(ListNode head) {
        if (head == null || head.next == null)
            return head;

        ListNode newHead = reverseRecursive(head.next);
        head.next.next = head;
        head.next = null;

        return newHead;
    }

    // Utility to print the linked list
    public static void printList(ListNode head) {
        while (head != null) {
            System.out.print(head.val + " -> ");
            head = head.next;
        }
        System.out.println("NULL");
    }

    public static void main(String[] args) {
        // Create Linked List: 1 → 2 → 3 → 4 → 5 → NULL
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        System.out.println("Original Linked List:");
        printList(head);

        // Reverse Iteratively
        ListNode reversedIter = reverseIterative(head);
        System.out.println("\nReversed Linked List (Iterative):");
        printList(reversedIter);

        // Reverse Recursively (to get original order back)
        ListNode reversedRec = reverseRecursive(reversedIter);
        System.out.println("\nReversed Linked List (Recursive):");
        printList(reversedRec);
    }
}

/*
 * ----------------------------------------------
 * Example Output:
 * Original Linked List:
 * 1 → 2 → 3 → 4 → 5 → NULL
 *
 * Reversed Linked List (Iterative):
 * 5 → 4 → 3 → 2 → 1 → NULL
 *
 * Reversed Linked List (Recursive):
 * 1 → 2 → 3 → 4 → 5 → NULL
 * ----------------------------------------------
 *
 * Time Complexity:
 * - Iterative: O(n)
 * - Recursive: O(n)
 *
 * Space Complexity:
 * - Iterative: O(1)
 * - Recursive: O(n)  (due to recursion call stack)
 */
